### ISSUE TITLE
Ubiquity Credit (uCR) Cash Out UI

### ISSUE THREAD

**Issue by 0x4007** [ubiquity/uad-ui#225]
We should make a simple UI that will allow a user to cash out uCR. This is intended to be the official "cash back rewards" interface. 

1. It will abstract the technical (conversion) details away from the end user.
2. It will ask the user if they want to redeem for uAD/USDC/DAI/USDT.
3. The UI will estimate the final result of the transactions e.g. 100 uCR -> 65 USDC.

### Above Peg
It will redeem uCR for uAD from our smart contract.

### Below Peg
It will market sell uCR for uAD.

### Finally
If the user requested another stablecoin, then it will market sell uAD for USDC/DAI/USDT.

**Comment by 0x4007** [comment]
This can go under the "Credits" view 

<img width="286" alt="image" src="https://user-images.githubusercontent.com/4975670/190066418-912bee22-b158-453b-84ad-43768ce87bda.png">

Perhaps the navbar can benefit from some icons next to each page title.

**Comment by 0xautohub** [comment]
Hey @pavlovcik I also want to finish this within a week alongside issue256(https://github.com/ubiquity/ubiquity-dollar/issues/256)

**Comment by 0x4007** [comment]
> Hey @pavlovcik I also want to finish this within a week alongside issue256(#256)

We generally only offer bounties in series to minimize blocking other developers to move things forward. Which do you prefer to work on first?

**Comment by 0x4007** [comment]
# Spec:

Don't work on CSS/aesthetics. I generally try to keep the CSS and classes as streamlined as possible and I'd rather handle this myself. Just make it functional! 

Inputs:

- [ ] uCR amount (with "max" button)
- [ ] Cash out asset (uAD/USDC/DAI/USDT) with the default set to uAD. 

The UI should estimate the output e.g. when the user selects USDC, then it should display "100 uCR -> 65 USDC" on the UI before they submit the transaction.

Todo: 

- [ ] make market for uAD/uCR on UniSwap (uCR liquidity should be basically worth 0.00 - 1.00 uAD)
- [ ] Deciding if we need ETH on the cash out list here still.

**Comment by 0xautohub** [comment]
I would like to work on Security CI(256) first and then move on this issue

**Comment by 0xautohub** [comment]
Hey @sergfeldman will you assign this issue to me?

**Comment by 0x4007** [comment]
> Hey @sergfeldman will you assign this issue to me?

> > Hey @pavlovcik I also want to finish this within a week alongside issue256(#256)
> 
> We generally only offer bounties in series to minimize blocking other developers to move things forward. Which do you prefer to work on first?

You will be assigned if this is available and the other task you were assigned is completed successfully

**Comment by codeawayaks** [comment]
You can assign this to me.

**Comment by 0xautohub** [comment]
Hey @sergfeldman I completed Security CI issue last week. want to work on this issue this week as I suggested before. Thanks

**Comment by sergfeldman** [comment]
@sunny0714 You are assigned to this issue.

**Comment by 0xautohub** [comment]
Cool @sergfeldman let me start

**Comment by 0xautohub** [comment]
Hey @pavlovcik In order to show the estimated exchange value(ex: 100 uCR -> 65 USDC) of token pairs(uCR/uAD,USDT,DAI,USDT), what api do you recommend? I have used Subgraph to get the exchange ratio for token pairs in one project before, but not sure at this issue
Any advice, please

**Comment by 0x4007** [comment]
I didn't do the research or have much experience with this sort of thing so I don't know the answer. I assume we need to still make some dedicated Uniswap market etc for uCR and other stables. Afterwards, I feel that the most direct way would be to query the smart contract directly for the ratios. Otherwise, I guess the Uniswap unofficial subgraph (I remember the official one was having problems a few months back)

**Comment by 0xautohub** [comment]
@pavlovcik thanks for your quick reply. Hope you to show me necessary links for querying the smart contract directly
Meanwhile, I would like to have a look at the Uniswap subgraph

**Comment by 0x4007** [comment]
It wouldn't be links, it would be custom code. For example, you get the ABI of the Uniswap pair, then you query the contract using Ethers etc.

**Comment by 0xautohub** [comment]
> 

So we need to wait to make dedicated Uniswap market for uCR, in order to get exact exchange ratio of those mentioned above?

**Comment by 0x4007** [comment]
No need to wait. Just confirm with me what pairs need to be made by when and I can handle it!

I'll start with the following:

```
uCR-uAD
uCR-DAI
uCR-USDC
uCR-USDT
```

**Comment by 0xautohub** [comment]
> No need to wait. Just confirm with me what pairs need to be made by when and I can handle it!
> 
> I'll start with the following:
> 
> ```
> uCR-uAD
> uCR-USDT
> uCR-DAI
> uCR-USDT
> ```

Cool, yeah right
One spell error: second pair: uCR-USDC

**Comment by 0xautohub** [comment]
Need to make right now, I have already completed this except for getting the ratio

**Comment by 0x4007** [comment]
All ready

**Comment by 0xautohub** [comment]
Hey @pavlovcik I have just opened PR for a review, please check this out

**Comment by sergfeldman** [comment]
@sunny0714 This bounty #225 is not fully completed at the moment. 
It is not possible to test it now and verify that it works as expected. 

The decision:

Alex created the new bounty #282 to provide what is missing
https://github.com/ubiquity/ubiquity-dollar/issues/282

Bounty #225 is dependent on #282
Both bounties will be paid after the full testing and completion

**Comment by 0xautohub** [comment]
> We should make a simple UI that will allow a user to cash out uCR. This is intended to be the official "cash back rewards" interface.
> 
> 1. It will abstract the technical (conversion) details away from the end user.
> 2. It will ask the user if they want to redeem for uAD/USDC/DAI/USDT.
> 3. The UI will estimate the final result of the transactions e.g. 100 uCR -> 65 USDC.
> 
> ### Above Peg
> It will redeem uCR for uAD from our smart contract.
> 
> ### Below Peg
> It will market sell uCR for uAD.
> 
> ### Finally
> If the user requested another stablecoin, then it will market sell uAD for USDC/DAI/USDT.

Hey @pavlovcik I have implemented this as follows
1. When below peg, allow users to swap uCR for uAD/USDC/USDT/DAI etc by using Uniswap swap widget
2. When above peg, allow users to swap uCR for uAD by using DebtCouponManager contract. And then swap uAD for USDC/USDT/DAI by using swap widget
 - when users try to swap uCR for uAD, the ui will only show the expected uAD amout by using Uniswap router implementation and will not show the swap widget
 - when users try to swap uCR for USDC/USDT/DAI, the ui will show both expected uAD and final token's amount by using router implementation. And also will show the swap widget to allow user to swap uAD for final token after getting uAD

The reasons I didn't implement the swap function by using Uniswap abis or API
1. There is no available API for swapping tokenA for tokenB
2. The quote amount of swap got from Router is depending on all pools(v2, v3), so if try to swap the quote amount, we should call swap functions on each pool, this might result in having most codes of swap widget in our code base. Then, our code might seem to look like sewed clothes

give me your opinion

**Comment by 0x4007** [comment]
> The reasons I didn't implement the swap function by using Uniswap abis or API
> 
> 1. There is no available API for swapping tokenA for tokenB
> 2. The quote amount of swap got from Router is depending on all pools(v2, v3), so if try to swap the quote amount, we should call swap functions on each pool, this might result in having most codes of swap widget in our code base. Then, our code might seem to look like sewed clothes

1. I found a code example below. As I mentioned, just calling the contract directly using Ethers should be straightforward.
2. I don't understand why we must call swap on every pool. We should only be performing two transactions: uCR redemption, and a uCR->XXX swap


```javascript
// Importing Dependencies
const { ethers } = require("ethers")
const UNISWAP = require("@uniswap/sdk")
const fs = require('fs');
const { Token, WETH, Fetcher, Route, Trade, TokenAmount, TradeType, Percent} = require("@uniswap/sdk");
const { getAddress } = require("ethers/lib/utils");

// Configure Infra Provider
const QUICKNODE_HTTP_ENDPOINT = "YOUR_QUICKNODE_HTTP_URL"
let provider = new ethers.providers.getDefaultProvider(QUICKNODE_HTTP_ENDPOINT)

// Import Wallet
const privateKey = fs.readFileSync(".secret").toString().trim()
const wallet = new ethers.Wallet(privateKey, provider)

// Instantiate Router Contract
UNISWAP_ROUTER_ADDRESS = "0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D"
UNISWAP_ROUTER_ABI = fs.readFileSync("./abis/router.json").toString()
UNISWAP_ROUTER_CONTRACT = new ethers.Contract(UNISWAP_ROUTER_ADDRESS, UNISWAP_ROUTER_ABI, provider)

// Import Token Data
const DAI = new Token(
    UNISWAP.ChainId.RINKEBY,
    "0x5592EC0cfb4dbc12D3aB100b257153436a1f0FEa",
    18
);
// https://www.quicknode.com/guides/defi/how-to-swap-tokens-on-uniswap-with-ethersjs
```

**Comment by 0xautohub** [comment]
> > The reasons I didn't implement the swap function by using Uniswap abis or API
> > 
> > 1. There is no available API for swapping tokenA for tokenB
> > 2. The quote amount of swap got from Router is depending on all pools(v2, v3), so if try to swap the quote amount, we should call swap functions on each pool, this might result in having most codes of swap widget in our code base. Then, our code might seem to look like sewed clothes
> 
> 1. I found a code example below. As I mentioned, just calling the contract directly using Ethers should be straightforward.
> 2. I don't understand why we must call swap on every pool. We should only be performing two transactions: uCR redemption, and a uCR->XXX swap
> 
> ```js
> // Importing Dependencies
> const { ethers } = require("ethers")
> const UNISWAP = require("@uniswap/sdk")
> const fs = require('fs');
> const { Token, WETH, Fetcher, Route, Trade, TokenAmount, TradeType, Percent} = require("@uniswap/sdk");
> const { getAddress } = require("ethers/lib/utils");
> 
> // Configure Infra Provider
> const QUICKNODE_HTTP_ENDPOINT = "YOUR_QUICKNODE_HTTP_URL"
> let provider = new ethers.providers.getDefaultProvider(QUICKNODE_HTTP_ENDPOINT)
> 
> // Import Wallet
> const privateKey = fs.readFileSync(".secret").toString().trim()
> const wallet = new ethers.Wallet(privateKey, provider)
> 
> // Instantiate Router Contract
> UNISWAP_ROUTER_ADDRESS = "0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D"
> UNISWAP_ROUTER_ABI = fs.readFileSync("./abis/router.json").toString()
> UNISWAP_ROUTER_CONTRACT = new ethers.Contract(UNISWAP_ROUTER_ADDRESS, UNISWAP_ROUTER_ABI, provider)
> 
> // Import Token Data
> const DAI = new Token(
>     UNISWAP.ChainId.RINKEBY,
>     "0x5592EC0cfb4dbc12D3aB100b257153436a1f0FEa",
>     18
> );
> // https://www.quicknode.com/guides/defi/how-to-swap-tokens-on-uniswap-with-ethersjs
> ```

Nice example, will look at it more carefully. Nice comment 👍

**Comment by 0xautohub** [comment]
> > > The reasons I didn't implement the swap function by using Uniswap abis or API
> > > 
> > > 1. There is no available API for swapping tokenA for tokenB
> > > 2. The quote amount of swap got from Router is depending on all pools(v2, v3), so if try to swap the quote amount, we should call swap functions on each pool, this might result in having most codes of swap widget in our code base. Then, our code might seem to look like sewed clothes
> > 
> > 
> > 
> > 1. I found a code example below. As I mentioned, just calling the contract directly using Ethers should be straightforward.
> > 2. I don't understand why we must call swap on every pool. We should only be performing two transactions: uCR redemption, and a uCR->XXX swap
> > 
> > ```js
> > // Importing Dependencies
> > const { ethers } = require("ethers")
> > const UNISWAP = require("@uniswap/sdk")
> > const fs = require('fs');
> > const { Token, WETH, Fetcher, Route, Trade, TokenAmount, TradeType, Percent} = require("@uniswap/sdk");
> > const { getAddress } = require("ethers/lib/utils");
> > 
> > // Configure Infra Provider
> > const QUICKNODE_HTTP_ENDPOINT = "YOUR_QUICKNODE_HTTP_URL"
> > let provider = new ethers.providers.getDefaultProvider(QUICKNODE_HTTP_ENDPOINT)
> > 
> > // Import Wallet
> > const privateKey = fs.readFileSync(".secret").toString().trim()
> > const wallet = new ethers.Wallet(privateKey, provider)
> > 
> > // Instantiate Router Contract
> > UNISWAP_ROUTER_ADDRESS = "0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D"
> > UNISWAP_ROUTER_ABI = fs.readFileSync("./abis/router.json").toString()
> > UNISWAP_ROUTER_CONTRACT = new ethers.Contract(UNISWAP_ROUTER_ADDRESS, UNISWAP_ROUTER_ABI, provider)
> > 
> > // Import Token Data
> > const DAI = new Token(
> >     UNISWAP.ChainId.RINKEBY,
> >     "0x5592EC0cfb4dbc12D3aB100b257153436a1f0FEa",
> >     18
> > );
> > // https://www.quicknode.com/guides/defi/how-to-swap-tokens-on-uniswap-with-ethersjs
> > ```
> 
> Nice example, will look at it more carefully. Nice comment 👍

Two transactions are uCR -> uAD, and uAD->XXX swap, right?

**Comment by 0xautohub** [comment]
Will you share the example link?

**Comment by 0x4007** [comment]
> Will you share the example link?

I don't understand what you're expecting. What's not clear?

**Comment by 0xautohub** [comment]
> > Will you share the example link?
> 
> I don't understand what you're expecting. What's not clear?

Not sure which function I should call to swap with router address related to pool or 2 tokens

**Comment by 0xautohub** [comment]
Guessing that Trade entity in uniswap/v3-sdk package can be used, but exact implementation example might help me a lot

**Comment by 0xautohub** [comment]
Directly calling the methods on router contract might work?

**Comment by 0xautohub** [comment]
I'm going to look at the Uniswap router contract and sdks carefully, if you @pavlovcik have a clear function call with params for swap, please update me

**Comment by 0xautohub** [comment]
@pavlovcik Implemented swap function when above peg by referencing Uniswap doc. But I couldn't test and only checked the code because it's a real transaction. Hope you to review the code and test it on your side

**Comment by 0x4007** [comment]
> @pavlovcik Implemented swap function when above peg by referencing Uniswap doc. But I couldn't test and only checked the code because it's a real transaction. Hope you to review the code and test it on your side

Like we discussed before, if you have blockers that cost gas fees; just ask for a budget approval ahead of time. I'll add $50 to your bounty to reimburse gas fees.

**Comment by 0xautohub** [comment]
> > @pavlovcik Implemented swap function when above peg by referencing Uniswap doc. But I couldn't test and only checked the code because it's a real transaction. Hope you to review the code and test it on your side
> 
> Like we discussed before, if you have blockers that cost gas fees; just ask for a budget approval ahead of time. I'll add $50 to your bounty to reimburse gas fees.

Okay. can you send some ETH for test now?

**Comment by 0x4007** [comment]
This would be paid in a single transaction, in USDC, upon the bounty completion. For now, sure I can send the ETH to your wallet address (please provide your address) but not again in the future as it makes accounting more complex than it needs to be.

**Comment by 0xautohub** [comment]
Yeah, understand, now I don't have any eth in my wallet. so.. :)
0xc6fa133f3290e14Ad91C7449f8D8101A6f894E25

**Comment by 0x4007** [comment]
https://etherscan.io/tx/0x7ff0e5fa2faf4770bcfc220a19a5927ee016eaf8970ce5e63ebffb3d0eb7f156

> Yeah, understand, now I don't have any eth in my wallet. so.. :) 0xc6fa133f3290e14Ad91C7449f8D8101A6f894E25

**Comment by 0xautohub** [comment]
@pavlovcik I can't redeem uCR for uAD because now below peg :(
![image](https://user-images.githubusercontent.com/5579669/196543939-87782eb4-9caf-4b66-97dd-c1e3d17bfee4.png)

**Comment by 0x4007** [comment]
> @pavlovcik I can't redeem uCR for uAD because now above peg :(

Let me manipulate the price... I need a few minutes

**Comment by 0x4007** [comment]
Unfortunately there is a blocking UI issue with crv.to so I need to figure out another UI for swapping on the metapools.

**Comment by 0xautohub** [comment]
> Unfortunately there is a blocking UI issue with crv.to so I need to figure out another UI for swapping on the metapools.

Oh.. sad story. hope it will be handled quickly

**Comment by 0x4007** [comment]
> > Unfortunately there is a blocking UI issue with crv.to so I need to figure out another UI for swapping on the metapools.
> 
> Oh.. sad story. hope it will be handled quickly

For now, unfortunately, the only other possibility to test now would be if you use Tenderly/Foundry to fork the network, see if you can impersonate a swap to the v2 Metapool (`0x20955CB69Ae1515962177D164dfC9522feef567E`) from ubq.eth to jack up the price on uAD. Then see if the uCR->uAD redemption works. If its too difficult to set up, we can pause this issue until we figure a UI workaround for crv.to or something.

I might also be able to extract the calldata from [an old buy transaction](https://etherscan.io/tx/0x452ffc25d8183643f51a88992e90ded64ad626cc660c49025a3d49a1dad7ae62)... will take me some time to research this though.

**Comment by 0x4007** [comment]
I think I generated my calldata ```0xa6417ed60000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000021e19e0c9bab240000000000000000000000000000000000000000000000000021674aed40640400000``` which should be 10k DAI -> (at least 9858.969226038560130000) uAD

Do you know a UI I can use to execute?

**Comment by 0xautohub** [comment]
> Do you know a UI I can use to execute?

Sorry, but I don't know

**Comment by 0x4007** [comment]
Injecting code in my console on etherscan which allows me to use 
```javascript 
await window.ethereum.request({
		method: "eth_sendTransaction",
		params: [transactionParameters],
	});
```

Swap simulation successful [on tenderly](https://dashboard.tenderly.co/pavlovcik/ubiquity-algorithmic-dollar/simulator/168ab552-af9b-43c7-888a-fdc53da693ab)

I asked "consulted" the TWAP oracle [on the simulation fork](https://dashboard.tenderly.co/pavlovcik/ubiquity-algorithmic-dollar/fork/cb34d553-222c-44ef-bfa1-5c79065f3e97/simulation/dbcd8147-18a6-4681-8e9a-90afe9e53272?trace=0) and it didn't update @zgorizzo69 how do I force update the TWAP?

<img width="373" alt="image" src="https://user-images.githubusercontent.com/4975670/196554017-a16c9fe7-1a05-4a0f-9cd7-499c302b115a.png">

**Comment by 0x4007** [comment]
@0xcodercrane could use your help for updating the TWAP on the simulation. I'm going through all the stack traces now and I'm stuck and about to end my work day.

**Comment by 0xautohub** [comment]
> @0xcodercrane could use your help for updating the TWAP on the simulation. I'm going through all the stack traces now and I'm stuck and about to end my work day.

Okay

**Comment by 0x4007** [comment]
I got word back on how to test @sunny0714 code. @0xcodercrane you think you can handle this today? 

@zgorizzo69 
> run a local fork call priceReset and connect the front end to the local fork

> 1- run ganache to fork mainnet 
> 2- run a small script to impersonate the deployer and call pricereset twice to restore the peg
> 3- connect metamask to your local node 
> 4- relax and test  sunny’s cash out rewards UI

**Comment by pavlovcik** [comment]
https://t.me/c/1588400061/524

**Comment by cuddle-dev** [comment]
yeah i will handle this stuff today

**Comment by sergfeldman** [comment]
This issue is waiting for testing.

**Comment by 0x4007** [comment]
Old pricing is still valid, just cleaning up some deprecated labels.

**Comment by 0xautohub** [comment]
Hi @pavlovcik this issue is ready for testing? If yes, which issues should I follow to test properly?
Or after team finish testing, I need to fix any issues generated through testing?

**Comment by 0x4007** [comment]
We still need to finish #295 in order to have a UI to swap for our dollar, moving the TWAP and allowing us to test easily.

**Comment by 0xautohub** [comment]
@pavlovcik how is the migration going? Testing is available?

**Comment by 0x4007** [comment]
Thanks for the follow up @sunny0714. The code for the migration seems to be ready but I still need to coordinate a time to work with @hashedMae and @zgorizzo69 to make sure I don't break something when trying to run it.

**Comment by 0xautohub** [comment]
Hey all, the work of this issue has been delivered and merged at that time, and it($1000) should be paid, but I have been told that you can pay for the work after full testing with migration to V3 Metapool.
It's quite unfair to me to release the bounty back without any payment.

I also have one another bounty([Enable Etherscan Support for Diamond](https://github.com/ubiquity/ubiquity-dollar/issues/322)) that you didn't pay and removed. Originally, it was completed and delivered at the current status of the repo, but I have asked to wait and update more after erc-2525 issue completed in all

Hope you guys to handle this properly
Thanks

**Comment by 0x4007** [comment]
@rndquu what do you think about unblocking and paying these out?

**Comment by rndquu** [comment]
> @rndquu what do you think about unblocking and paying these out?

Regarding https://github.com/ubiquity/ubiquity-dollar/issues/322

The corresponding PR: https://github.com/ubiquity/ubiquity-dollar/pull/407

@sunny0714 

What this [PR](https://github.com/ubiquity/ubiquity-dollar/pull/407) implements:
1. Deploys a dummy implementation of the Diamond contract 

How it should've been implemented:
1. Deploy a dummy implementation of the Diamond contract
2. Add the `EtherscanFacet` to the Diamond
3. Set the proxy to point to the deployed dummy implementation from step 1

So the issue is not solved

---

Regarding https://github.com/ubiquity/ubiquity-dollar/issues/225

The corresponding [PR](https://github.com/ubiquity/ubiquity-dollar/pull/279) actually implements the spec at a bare minimum. 

For such bounty price the implementation must be almost perfect, I mean: 
- [properly tested](https://github.com/ubiquity/ubiquity-dollar/issues/225#issuecomment-1282915540)
- with error handling
- without [double approvals](https://github.com/ubiquity/ubiquity-dollar/pull/279/files#diff-f3bc89a3c4a369c10dec26337298b9d4c1c2a80fea7671796912a0252641ae3bR47)
- with corresponding code comments of what the hook or UI component does
- unit testing

Right now we have a PR with an obsolete code which we are able to use only as a reference.

@pavlovcik  I would make a payout but reduce the bounty price by 50%

**Comment by 0xautohub** [comment]
> How it should've been implemented:
> 
> 1. Deploy a dummy implementation of the Diamond contract
> 2. Add the `EtherscanFacet` to the Diamond
> 3. Set the proxy to point to the deployed dummy implementation from step 1

Originally it was implemented in the PR, but it was cancelled when syncing the main repo

**Comment by rndquu** [comment]
> > How it should've been implemented:
> > 
> > 1. Deploy a dummy implementation of the Diamond contract
> > 2. Add the `EtherscanFacet` to the Diamond
> > 3. Set the proxy to point to the deployed dummy implementation from step 1
> 
> Originally it was implemented in the PR, but it was cancelled when syncing the main repo

I don't see any commits in this [PR](https://github.com/ubiquity/ubiquity-dollar/pull/407)

**Comment by 0xautohub** [comment]
> I don't see any commits in this [PR](https://github.com/ubiquity/ubiquity-dollar/pull/407)

Yeah, all commits cancelled

So I said that I don't want to get paid for that issue, and willing to continue to work if more details are provided like goerli testnet deployment

**Comment by 0xautohub** [comment]
> @pavlovcik I would make a payout but reduce the bounty price by 50%

@rndquu It's too tight to suggest to reduce the bounty price by 50% with small bits of drawbacks you found. I would like you to pay $1000 in all for 2 issues, if it's not quite acceptable, just suggestion

**Comment by 0xautohub** [comment]
> * [properly tested](https://github.com/ubiquity/ubiquity-dollar/issues/225#issuecomment-1282915540)
> * with error handling

It can't be tested on my side, as you know


> * without [double approvals](https://github.com/ubiquity/ubiquity-dollar/pull/279/files#diff-f3bc89a3c4a369c10dec26337298b9d4c1c2a80fea7671796912a0252641ae3bR47)

What's double approvals?

> * with corresponding code comments of what the hook or UI component does

This can be added

> * unit testing

I'm sorry, but I can't find any unit testing for frontend functions and unit testing hasn't been asked in the issue description

**Comment by 0xautohub** [comment]
@pavlovcik  please define this bounty price again and settle it today. I've already experienced that It's quite difficult to get full amount of the bounty after delivering while working on erc-2535 issue.

Here is my address
0xe50ae2abfe7a36f934b71f0e3f96012e47951132

Best regards
Thanks
Sunny

**Comment by sunny0714** [comment]
OMG, my TG account has just been hacked :(

**Comment by 0xautohub** [comment]
@pavlovcik @rndquu Could I get updates now?

**Comment by 0xautohub** [comment]
When can I get updates?

**Comment by 0x4007** [comment]
I'm leaning heavily on @rndquu's opinion because 1. I am not a solidity developer and 2. I do not have the ability to deep dive into GitHub this week. 

However I spent focused time reviewing this now. 

This is certainly an exceptional situation, with a series of multiple unfortunate events like crv.to UI going down through the project duration, and the cascade effect of blockers to proceed on any of the active tasks you were working on. 

Your contributions to the project are appreciated and we want to thank you for working with us when we were first figuring things out with our bounty system. I think we learned a lot about how to structure our bounties from these experiences. 

I don't anticipate this sort of situation happening again so we will settle for the full amount as long as you acknowledge the following:

1. If the tasks are not implemented according to their specifications, then they are not eligible to receive payments for the bounties. This means that #407 is not eligible for the bounty payment. 
2. There seems to be ongoing issues that are not your responsibility but are blocking these from completion. This means that I would not encourage you to pick up these same tasks to try and complete them at this time. 
3. We can not make use of your deliverables, aside from using #225 as a point of reference for reimplementation (which we will need to pay again for.)
4. I believe that we learned from these early bounties and now when we create tasks they are far less likely to lead to these outcomes. Thanks for the opportunity to be able to learn and improve from.  

I think ultimately it's worth considering that we are paying out of principle, and as a thank you for your patience and early contributions. This is not a payment because we received what we needed from the original agreement. 

This task system isn't perfect but moments like these help us improve processes to make things more robust. Thank you for the opportunity for improvement. 

https://etherscan.io/tx/0x0ea554204f52a23c38e884b302099b69b8e26f9c2b3c537d8a36d18a9e346263
